engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -3706,6 +3691,34 @@ public void checkIfCanUpgrade(final VirtualMachine vmInstance, final ServiceOffe
         }
     }
 
+    protected void checkIfNewOfferingStorageScopeMatchesStoragePool(VirtualMachine vmInstance, ServiceOffering newServiceOffering) {
+        boolean isRootVolumeOnLocalStorage = isRootVolumeOnLocalStorage(vmInstance.getId());
+
+        if (newServiceOffering.isUseLocalStorage() && !isRootVolumeOnLocalStorage) {
+            String message = String .format("Unable to upgrade virtual machine %s, target offering use local storage but the storage pool where "
+                    + "the volume is allocated is a shared storage.", vmInstance.toString());
+            throw new InvalidParameterValueException(message);
+        } else if (!newServiceOffering.isUseLocalStorage() && isRootVolumeOnLocalStorage) {
```

Comment:
else is not needed here and might cause confusion
```suggestion
        } 
        if (!newServiceOffering.isUseLocalStorage() && isRootVolumeOnLocalStorage) {
```

Reply:
Thanks for the suggestion @DaanHoogland, makes sense :+1: